### DISCUSSION TITLE
Merge Two Datasets

### DISCUSSION THREAD

**Discussion by khaliduzzaman-mredul** [ultralytics/ultralytics#10075] [General]
Recently I was working on a project with 20,000 images and I used roboflow to label the images. Since roboflow has a limitation of 10,000 images I labeled the images using two different accounts and then exported the datasets as YOLOv8 format and then I merged the images in a single folder accordingly. But after the training it appears that although the images and annotations were merged but only images from the first account was trained and the second one was skipped.

Can anyone help me to find out how to properly merge the two datasets.
Please note that I also modified the data.yaml file.

**Comment by pderrenger** [comment]
@SERPENT47 it sounds like you've put great effort into labeling your datasets. Merging datasets can sometimes lead to issues if not done carefully. Here are a couple of things you might want to check:

1. **Overlapping Filenames:** Ensure there's no overlap in image or annotation filenames between the two datasets. If they share names, one could overwrite the other during the merge.

2. **data.yaml Configuration:** Double-check that the `data.yaml` file correctly points to all merged images and annotations. It should list paths that collectively include files from both original datasets.

3. **Format Consistency:** Verify that the YOLOv8 format export from Roboflow did not differ between the two accounts for any reason, affecting the usability of one dataset over the other.

As an example, your `data.yaml` might look something like this, ensuring it covers both subsets:

```yaml
train: ./data/images/train/
val: ./data/images/val/

...

# Classes
nc: 80  # Number of classes
names: ['class1', 'class2', ...]  # Class names
```

Make sure the `train` directory contains all the training images and annotations from both accounts, and similar for `val`. If you've done all this and still face issues, it might help to look at the loading part of your training logs to ensure files from both datasets are being recognized and loaded.

Keep experimenting and don't hesitate to reach out if you have more questions! 🙌

**Reply by khaliduzzaman-mredul** [reply]
Hey thanks for the reply. But unfortunately I did everything exactly as your instructions already. At the start of the training it also loads all the 16,000 images [rest 4000 were for test and validation] 
But it only trains images from the first account and completely ignores all the classes from the 2nd one.

**Reply by pderrenger** [reply]
Hi there! 😊 It seems like you've made sure everything is set up correctly, which is great! Since it recognizes all 16,000 images at the start, the issue might relate to class annotations in your datasets. Could there be a mismatch or difference in class names or class IDs between the datasets from both accounts? Please ensure that the class names and their order in `data.yaml` accurately reflect both datasets. A snippet showing how classes are defined might look like this:

```yaml
nc: number_of_classes
names: ['class1', 'class2', ..., 'classN']
```

This part is crucial for successful training across merged datasets. If everything appears correct, checking for any warnings or errors related to annotations during training startup might also shed light on the problem. Keep pushing, and let's get those classes recognized! 👍

**Reply by khaliduzzaman-mredul** [reply]
The number of classes and names were also merged properly. I guess I shall check the order again and will let you know the update.

**Reply by pderrenger** [reply]
@SERPENT47 hey there! 👋 Absolutely, double-checking the order sounds like a good plan. It's crucial that the class names in your `data.yaml` reflect the exact ordering and names you used across both datasets. Sometimes, it's the little details that trip us up! If everything matches and you're still facing issues, feel free to drop another message. Looking forward to hearing back from you! 🚀

**Reply by khaliduzzaman-mredul** [reply]
Well I checked the order again but unfortunately didn't make any difference!

**Reply by pderrenger** [reply]
@SERPENT47 hey there! 😊 Sorry to hear that the issue persists. If the order and class names are correctly merged, let's quickly verify a few more aspects:

- Ensure no two classes share the same name across the datasets.
- Check for potential spelling or case sensitivity issues in class names within `data.yaml` and annotation files.

If everything checks out and you're still facing problems, could you share a snippet of your `data.yaml` and perhaps an example annotation from each dataset? This might help pinpoint the discrepancy. Keep up the great work!

**Comment by khaliduzzaman-mredul** [comment]
Here is the preview of my data.yaml file:

`train: ../train/images
val: ../valid/images
test: ../test/images

nc: 39
names: ['ananda_tv', 'asian_tv', 'atn_bangla', 'atn_news', 'bangla_tv', 'banglavision', 'bijoy_tv', 'boishakhi_tv', 'btv', 'btv_chattogram', 'btv_sangsad', 'btv_world', 'channel_24', 'channel_9', 'channel_i', 'dbc_news', 'deepto_tv', 'desh_tv', 'duronto_tv', 'ekattor_tv', 'ekhon', 'ekushey_tv', 'gaan_bangla', 'gazi_tv', 'global_tv', 'green_tv', 'independent_tv', 'jamuna_tv', 'maasranga_tv', 'mohona_tv', 'my_tv', 'n_tv', 'nagorik_tv', 'news_24', 'nexus_tv', 'r_tv', 'sa_tv', 'somoy_tv', 't_sports']`

And here you will find the annotations: https://drive.google.com/drive/folders/1YRK7e6r1GV_ynqUHdxspzJQEB9Jhv3zH?usp=drive_link

**Reply by glenn-jocher** [reply]
Hey there 👋! It looks like you've done a great job setting up your `data.yaml` file and organizing your dataset. If you've already merged the image and annotation files from both accounts into the same directories (`../train/images`, `../valid/images`, etc.), and updated your `data.yaml` to reflect the total classes and class names accurately, you're mostly there.

To troubleshoot the issue of the second dataset apparently not being used during training, consider the following checks:

- **Ensure no name conflicts:** Make sure there are no identical filenames in the first and second datasets. If there are, rename them to prevent overwriting.
- **Verify annotations are correct:** Double-check a few annotation files from the second dataset manually to ensure they're correctly formatted for YOLOv8 and point to the right class indices.
- **Look for errors in the logs:** When you start the training, carefully review the console output for any warnings or errors related to loading images or annotations.
- **Check image paths:** If your datasets came from different accounts, ensure the paths in the annotation files correctly match where you've stored the images after merging.

If everything seems to be in order, try running the training again while closely monitoring the initial log messages for any hints on what might be going wrong.

Lastly, sometimes it's a good idea to do a "dry run" of a small subset of the merged dataset (including a few images and annotations from both original datasets). This can help identify any structured issues before committing to a long training process.

If after trying these steps, the issue persists, feel free to share more details or error messages if any. Happy training! 🚀

**Reply by khaliduzzaman-mredul** [reply]
Well at this point I am completely sure that everything you mentioned are properly done. For testing purposes I took just one image for the 39 classes so that there is no file name related issues between the images and label files and also I thoroughly checked the data.yaml file which provided earlier.
But yet no progress. Still it's taking the first 20 classes which was from the 1st account for training.

If you want to give it a try yourself then I provided the data.yaml file and drive link for the dataset folder in my previous comment.

**Reply by glenn-jocher** [reply]
Hey there! 😊 It sounds like you've been thorough with your checks. Given your setup, one possibility might be that the model is not recognizing the added classes due to caching. Try clearing any cache files (usually named something like `.cache` or similar) in your dataset folder and initiate training anew. If your dataset structure and `data.yaml` are correctly set, this should ideally solve the issue.

Remember, the training process scans and caches dataset details early on for efficiency. If there were any changes after the initial cache creation (like your dataset modifications), clearing this cache is necessary.

Here's a quick command to remove cache files if they're in your dataset directory:
```bash
rm -f ../train/*.cache ../valid/*.cache ../test/*.cache
```

Then, initiate your training again. Fingers crossed this gets things moving! 🤞

**Reply by khaliduzzaman-mredul** [reply]
Yet no luck!

**Reply by glenn-jocher** [reply]
Hey there! 😊 I'm sorry to hear you're still facing issues. Could you please check if there's a `data.yaml` copy inside your `train`, `valid`, and `test` folders? If yes, remove them, ensuring only the main `data.yaml` in your project root is present. Also, try clearing any potential cache with:

```bash
find . -type f -name '*.cache' -delete
```

Then, give training another go. If the issue persists, it might be helpful to run a single training step/debug mode to get more insights. If you need further assistance, feel free to share!

**Reply by khaliduzzaman-mredul** [reply]
All these were checked but no luck. 
At this point I think there is something from the  Roboflow side which is interfering with the process.

**Reply by glenn-jocher** [reply]
Hello! Thanks for reaching out. It's possible that the issue might be related to how the data was processed or exported from Roboflow. A common pitfall to check for is ensuring that all your classes are correctly labeled across both datasets and consistently represented in the `data.yaml` file.

Also, when merging, make sure that your dataset annotations (in both sets) are indeed in the YOLOv8 format and that no class indices are mismatched or omitted. A quick way to verify is to open a few label files from each dataset and manually check if the class indices and bounding box coordinates make sense.

If you've confirmed everything is correct on your end, it could be beneficial to reach out to Roboflow support with a detailed description of how you processed and exported your datasets. They might be able to provide more insight or point out something that might have been missed.

Keep up the great work, and I hope this gets resolved soon! 😊
